mobius-core/src/test/java/com/spotify/mobius/MobiusLoopObservabilityBehavior.java

```diff
@@ -55,4 +68,68 @@ public void shouldSupportUnregisteringObserver() throws Exception {
 
     observer.assertStates("init", "init->active observer");
   }
+
+  @Test
+  public void shouldNotReportModelsInIncorrectOrder() throws Exception {
+    // 1. create a loop with initial model A
+    // 2. concurrently, do:
+    //    a. dispatch an event that changes the model to B
+    //    b. add an observer
+    // 3. verify that the observer never saw B, A.
+
+    ExecutorService service1 = Executors.newSingleThreadExecutor();
+    ExecutorService service2 = Executors.newSingleThreadExecutor();
+
+    List<String> bad = new ArrayList<>();
+
+    // 100,000 iterations tended to lead to about 10 instances of 'B,A' before the underlying
+    // issue was fixed - but that takes a lot of time for little benefit now that the issue is
+    // fixed,  so this iteration count was reduced. If another out-of-order error exists, or the
+    // error happens again, this test should become flaky.
+    for (int i = 0; i < 1000; i++) {
+      MobiusLoop<Integer, Integer, Integer> loop =
+          MobiusLoop.create(
+              (model, event) -> Next.next(event),
+              0,
+              emptyList(),
+              new NoopConnectable(),
+              new NoopConnectable(),
+              new ImmediateWorkRunner(),
+              new ImmediateWorkRunner());
+
+      List<Integer> observed = Collections.synchronizedList(new ArrayList<>());
+
+      final Future<Disposable> future2 = service2.submit(() -> loop.observe(observed::add));
+      final Future<?> future1 = service1.submit(() -> loop.dispatchEvent(1));
+
+      future1.get();
+      future2.get();
+
+      assertThat(observed.size()).isBetween(1, 2);
+
+      // two models, with model 0 at the end is bad - adding that to a list so that the test will
+      // give a clearer indication of how common this was.
+      if (observed.size() == 2 && observed.get(1) == 0) {
+        bad.add(observed.toString());
+      }
+    }
+
+    assertThat(bad).isEmpty();
+  }
+
+  private static class NoopConnectable implements Connectable<Integer, Integer> {
```

Comment:
hmmm I think I've seen NoopConnectables in some client code before, maybe we should consider adding one to the library?